src/components/checkbox/checkbox.ts

```diff
@@ -0,0 +1,246 @@
+ */
+@Component({
+  selector: 'md-checkbox',
+  templateUrl: './components/checkbox/checkbox.html',
+  styleUrls: ['./components/checkbox/checkbox.css'],
+  host: {
+    '[id]': 'id',
+    '[class.md-checkbox]': 'true',
+    '[class.md-checkbox-indeterminate]': 'indeterminate',
+    '[class.md-checkbox-checked]': 'checked',
+    '[class.md-checkbox-disabled]': 'disabled',
+    '[class.md-checkbox-align-end]': 'align == "end"',
+    '[attr.role]': '"checkbox"',
+    '[attr.tabindex]': 'disabled ? -1 : tabindex',
+    '[attr.aria-label]': 'ariaLabel',
+    '[attr.aria-labelledby]': 'labelId',
+    '[attr.aria-checked]': 'getAriaChecked()',
+    '[attr.aria-disabled]': 'disabled',
+    '(click)': 'toggle($event)',
+    '(keyup.space)': 'toggle($event)',
+    '(blur)': 'onTouched()'
+  },
+  providers: [MD_CHECKBOX_CONTROL_VALUE_ACCESSOR],
+  encapsulation: ViewEncapsulation.None
+})
+export class MdCheckbox implements ControlValueAccessor {
+  /**
+   * Attached to the aria-label attribute of the host element. In most cases, arial-labelledby will
+   * take precedence so this may be omitted.
+   */
+  @Input('aria-label') ariaLabel: string = '';
+
+  /** A unique id for the checkbox. If one is not supplied, it is auto-generated. */
+  @Input() id: string = `md-checkbox-${++nextId}`;
+
+  /** Whether or not the checkbox should come before or after the label */
+  @Input() @OneOf(['start', 'end']) align: string = 'start';
+
+  /**
+   * True when the checkbox is disabled, false otherwise. When the checkbox is disabled it cannot
+   * be interacted with. The correct ARIA attributes are applied to denote this to assistive
+   * technology.
+   */
+  @Input() disabled: boolean = false;
+
+  /**
+   * The tabindex attribute for the checkbox. Note that when the checkbox is disabled, the attribute
+   * on the host element will be set to -1, regardless of the actual tabindex value.
+   */
+  @Input() tabindex: number = 0;
+
+  /** Event emitted when the checkbox's `checked` value changes. */
+  @Output() change: EventEmitter<boolean> = new EventEmitter();
+
+  /** Called when the checkbox is blurred. Needed to properly implement ControlValueAccessor. */
+  onTouched: () => any = () => {};
+
+  private _currentAnimationClass: string = '';
+
+  private _currentUiState: UiTransitionState = UiTransitionState.Init;
+
+  private _checked: boolean = false;
+
+  private _indeterminate: boolean = false;
+
+  private _changeSubscription: {unsubscribe: () => any} = null;
+
+  constructor(private _renderer: Renderer, private _elementRef: ElementRef) {}
+
+  /**
+   * True when the checkbox is checked, false otherwise. Note that setting `checked` while
+   * will immediately set `indeterminate` to false.
+   */
+  @Input() get checked() {
+    return this._checked;
+  }
+
+  set checked(checked: boolean) {
+    // When checked is set, it immediately overwrites an indeterminate state, regardless of
+    // whether it's true or false.
+    this._indeterminate = false;
```

Comment:
It's tricky because the imperative API and the declarative API are at odds. The problem with having one of the properties un-set the other is that now the state of the component can depend on the order in which Angular sets the properties when you're using `[checked]` and `[indetermindate]`. 

My initial thought was to follow the native element completely, which makes it easy to defend the choice; "Of course we do it that way, it's what the native checkbox does." The more I think about it, though, the more I think the current approach is actually better since it more closely captures how people will actually want to use it.